### PR TITLE
Tonfa updates

### PR DESCRIPTION
**Pull request by PowerfulBacon**
All in all, I realise that making unique weapons without combat mode is near impossible, so this is the best that you get for now.

## About The Pull Request

Changes several aspects about the tonfa, simplifies a lot of the behaviour (since it was unnecessarilly complex) and tries to keep some unique behaviours.

- Removes the knock-out
- Removes the weird and confusing variable damage code, it is very difficult to reason about something that changes its damage depending on the target's health.
- Significant code improvements through simplifications.
- Aiming for legs will knockdown targets with enough stamina damage, and will push them back to give them some time to recover and prevent chain attacks.
- Fixes armour code using the wrong damage type.
- Tonfa will now deal 5 damage on help mode. This is relatively minor considering the stamina damage and is still non-lethal, but if you are going to be using physical force against a target then they are going to need to get a bit of medical attention afterwards.
- Tonfa is now the same size as the baton

Unarmoured:

|Bodyzone|Stun baton|Tonfa|
|-|-|-|
|Chest/Head|3 hit crit, 0/24 brute help/harm, knockback|4 hit crit, 20/48 brute help/harm|
|Legs|3 hit disable|4 hit knockdown, 5 hit disable|
|Arms|2 hit disable|4 hit disable|

Security Armour:

|Bodyzone|Stun baton|Tonfa|
|-|-|-|
|Chest/Head|4 hit crit|7 hit crit|
|Legs|3 hit disable|4 hit knockdown, 5-7 hit disable|
|Arms|2 hit disable|4 hit disable|

Captain Armour:

|Bodyzone|Stun baton|Tonfa|
|-|-|-|
|Chest/Head|5 hit crit|8 hit crit|
|Legs|2 hit disable|4 hit knockdown, 6-7 hit disable|
|Arms|3 hit disable|4 hit disable|

## Why It's Good For The Game

The previous tonfa used bodyzone targeting code incorrectly, had some very complicated logic for what should be simple behaviours and had some things that people didn't find fun. I'm not a fan of having more security weapons that need balancing, but at least there are some unique properties to justify its existence.

The tonfa is a stun baton with EMP immunity, so it should be slightly worse than it while still having at least some unique element to give it a reason to exist.

## Testing Photographs and Procedure

image

See testing tables above for results.

## Changelog
:cl:
code: Significant code improvements on the tonfa.
balance: Tonfa no longer causes unconciousness or paralysis, instead using regular stamina damage.
balance: Tonfa's damage has been standardised to make it easier to balance.
balance: Tonfa leg attack causes knockdown and pushback if the target has enough stamina damage.
balance: Tonfa is now the same size as the baton
/:cl:

### CODE REVIEW

File: code/game/objects/items/melee/misc.dm
Comment (PowerfulBacon) [inline code/game/objects/items/melee/misc.dm]:
```suggestion
	var/non_harm_force = 4
```
You know what, lets bring this down a bit

**Comment by TsunamiAnt** [comment]
Why the pushback to prevent chains? I thought the whole reason for the tonfa was comboing hitchains?

**Comment by PowerfulBacon** [comment]
There is nothing special about the tonfa for combing or hit-chaining, it is a standard melee weapon by all accounts. The push-back is to prevent the knock-down being an immediate win by giving the other person a short chance before the next attack, and to give them the advantage of being in the defending position.

**Comment by TsunamiAnt** [comment]
huh. fair enough. Is the knockdown longer then? Because it used to be just like a small stumble. Otherwise you're kinda hitting your enemy to give them an advantage.

**Comment by PowerfulBacon** [comment]
Knockdown ranges from 0 to 4 seconds between 22 and 72 damage.

**Comment by Fronsis** [comment]
Would it be possible to include a short video showcasing the knockdown-push back of the Tonfa alongside the action of critting someone? The rest seems fair enough to have it on a table, I recall you've shown recordings for testing evidence in the past like in your slowdown PR's. You can't really give a good feedback or to confirm the testing evidence solely based on a screenshot that show some naked spacemen, i, and i'm sure others, would greatly appreciate a video for testing evidence.

As for the rest it seems these are good balancing changes for the Tonfa

That's all thank you.

**Comment by PowerfulBacon** [comment]
Ah yea, my laptop is on windows 11 so PRs on that always have video since it's built into snipping tool but anything on my PC is pictures.

**Comment by Rukofamicom** [comment]
Would you consider making tonfa the same size as batons? This was one of the bigger upsides it had alongside emp immunity - they take up half the size of a baton.

**Comment by DarnTheMarn** [comment]
It should also take up the same slot on the sec belt as the stun baton so you can't carry two baton type items on you at a time just saying.

**Comment by TsunamiAnt** [comment]
MERGE IT ALREADY!!!!

**Review by Rukofamicom — APPROVED** [review]
I would still like to see the Tonfa as a large sized item, but not enough to block this.

File: code/game/objects/items/melee/misc.dm
Comment (EvilDragonfiend) [inline code/game/objects/items/melee/misc.dm:965]: This will tell different force value on examining everytime it's used with different intent.

File: code/game/objects/items/melee/misc.dm
Comment (EvilDragonfiend) [inline code/game/objects/items/melee/misc.dm]: I think It looks it should be called when it's early returned from other lines.

File: code/game/objects/items/melee/misc.dm
Comment (EvilDragonfiend) [inline code/game/objects/items/melee/misc.dm:993]: line 930 has a target zone. should this line exist?

File: code/game/objects/items/melee/misc.dm
Comment (EvilDragonfiend) [inline code/game/objects/items/melee/misc.dm]: I think these should be logged even if an attacker fails to attack.

File: code/game/objects/items/melee/misc.dm
Comment (EvilDragonfiend) [inline code/game/objects/items/melee/misc.dm:981]: this var name sounds like effectiveness to stamina attack. would you rename it a little bit? like `knuckdown_eff`

File: code/game/objects/items/melee/misc.dm
Comment (EvilDragonfiend) [inline code/game/objects/items/melee/misc.dm:940]: Hmm, I am not sure, but I wonder if these should stay here, or moved to below since `target_zone` would be only available from carbon mobs... but `armour_level`'s proc wants to have a target_zone...

File: code/game/objects/items/melee/misc.dm
Comment (EvilDragonfiend) [inline code/game/objects/items/melee/misc.dm:953]: line 951-954 adjusts `force` value, but it uses the force variable before changing it.

File: code/game/objects/items/melee/misc.dm
Comment (EvilDragonfiend) [inline code/game/objects/items/melee/misc.dm]: line 933 already adds fingerprint

**Review by EvilDragonfiend — CHANGES_REQUESTED** [review]

File: code/game/objects/items/melee/misc.dm
Comment (PowerfulBacon) [reply code/game/objects/items/melee/misc.dm:965]: Is that a problem? It's going to deal different force depending on your intent

File: code/game/objects/items/melee/misc.dm
Comment (PowerfulBacon) [reply code/game/objects/items/melee/misc.dm]: most other early returns aren't actually performing an attack, this exists for slime shitcode that I am going to make another PR removing

File: code/game/objects/items/melee/misc.dm
Comment (PowerfulBacon) [reply code/game/objects/items/melee/misc.dm:993]: Technically it would work in this case, however this is the correct way to check if a specific zone is being targetted as it handles edge cases. The edge cases don't apply here, but this is still the way to check to see if a specific zone should be considered as a target.

File: code/game/objects/items/melee/misc.dm
Comment (PowerfulBacon) [reply code/game/objects/items/melee/misc.dm:981]: I don't mean to be combative in response to a review, however I don't see how anyone could be confused by this variable when its only uses are a few lines down. The name effectiveness is pretty clear in indicating the general idea of what this variable does, and if you ctrl-f you can see that it is being used for knockdown. Self-documenting code is important, but there is a point where it becomes pointless, even going so far as to making the code more confusing. It is called effectiveness as it is a proportional ratio between 0 and 1, which I think is a pretty standard and perfectly fine variable name.

File: code/game/objects/items/melee/misc.dm
Comment (PowerfulBacon) [reply code/game/objects/items/melee/misc.dm:940]: Mobs that don't have bodyzones handle that inside of getarmor and will work perfectly fine. There is no issue with this code and this is the standard method for doing this.

File: code/game/objects/items/melee/misc.dm
Comment (PowerfulBacon) [reply code/game/objects/items/melee/misc.dm:965]: Actually I do see the issue, force should be recalculated when the force string is returned. Should we refactor force into a more extensible system?

File: code/game/objects/items/melee/misc.dm
Comment (PowerfulBacon) [reply code/game/objects/items/melee/misc.dm:953]: Good catch, resolved

File: code/game/objects/items/melee/misc.dm
Comment (PowerfulBacon) [reply code/game/objects/items/melee/misc.dm]: Good catch, this issue applies to all types of batons

File: code/game/objects/items/melee/misc.dm
Comment (EvilDragonfiend) [reply code/game/objects/items/melee/misc.dm:981]: fair enough

File: code/game/objects/items/melee/misc.dm
Comment (EvilDragonfiend) [reply code/game/objects/items/melee/misc.dm:965]: honestly, yes, we should if we can.
`initial(force)` isn't a good practice for most items.
For example, you can trim blades to increase its force damage. Increased force damage will be removed if `initial(force)` is used, or directly takes an adjusted value.

image

(Chainsaw code)
As you can see, damage isn't reliable.

hopefully there aren't much codes to manage force damage dynamically, but if you want to add something dynamic that change force damage, you'll need to consider something on this.

File: code/game/objects/items/melee/misc.dm
Comment (EvilDragonfiend) [reply code/game/objects/items/melee/misc.dm]: I see the PR.

**Review by EvilDragonfiend — APPROVED** [review]
Code looks alright, but I think Bacon should merge this himself when he thinks it's good to go.

**Comment by Nakalon** [comment]
It took me a long while to make this weapon.
I made it with the little knowledge I had available and I did the best I could.

I don't know if the result of my work reflected my goals or intentions.
When I made the Tonfa I didn't want it to be a suitable alternative to the stun baton.

It was meant to be a Roleplayers weapon.
You couldn't use the billy club in combat because you'd be bwoinked.
And at the time the stun batton was a 2 hit stam crit, so that was boring.
I wanted to create something dynamic and fun.

Hits to the legs would sometimes trip the target, but it wasnt anything useful, it was like half a second or so accompanied by a banana peel sound effect.
Hits to the arms would make people drop their weapon as other batons.
Hits to the torso would be the most efficient way to stam crit a target.
The final hit to the head was meant to be a secondary tool that could be used as a flash, giving enough time to handcuff the enemy.
I never understood what people had against the sleep that they don't have against the flash.
It at least has a funny sound to it, and it lasts no longer than the alternative.

I remember at that time, I had decided to make the weapon rather weak. But I was told otherwise. I was told to increase the damage, which was something I didn't really want to do to begin with.

Its all well and good with cleaning up my awful code and all that, but I think this basically kills what I had envisioned for the weapon.

It becomes another boring tool. Another Stun baton, which was what no one wanted it to become.

I know you know better than me, and that I'm probably wrong and speaking on behalf of my idea.

But please, please consider just cleaning the code, making the weapon bigger and nerfing the damage, but keeping the weapon and its unique skills as they are.

This was meant to be a flavor weapon, for RP, for people who enjoy that 1v1 with the unruly bar patron armed with a stool.

Please please consider my words, even if you decide to go against them.
Cheers!

**Comment by Nakalon** [comment]
https://github.com/BeeStation/BeeStation-Hornet/pull/10982

Whoever made this did not understand why I made the tonfa AT ALL.

"Using the tonfa is a nightmare. I understand it is intended to be used as a backup currently, but it doesn't do enough stamina damage to knock someone out usually requiring upwards of 6-7 hits."

No it was not. At all. As I've stated previously.

That high amount of hits was ideal, and I wanted it to be like 10 at the time.

I beg, revert these damage changes, see how the meta shifts for a week or so and make a final decision.

File: code/game/objects/items/melee/misc.dm
Comment (PowerfulBacon) [inline code/game/objects/items/melee/misc.dm]:
```suggestion
	var/non_harm_force = 3
```

File: code/game/objects/items/melee/misc.dm
Comment (Tsar-Salat) [inline code/game/objects/items/melee/misc.dm]: spelling

File: code/game/objects/items/shields.dm
Comment (Tsar-Salat) [inline code/game/objects/items/shields.dm]: should probably include /obj/item/melee/classic_baton subtypes too, no? 

It should clunk if I hit it with a tonfa, stunbaton, Billy club, or det baton

File: code/game/objects/items/melee/misc.dm
Comment (Tsar-Salat) [inline code/game/objects/items/melee/misc.dm]: spelling

File: code/game/objects/items/shields.dm
Comment (PowerfulBacon) [reply code/game/objects/items/shields.dm]: I'll let it work for any melee weapon that isn't sharp

**Pull request merged**
